clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -332,11 +333,53 @@ public int splitAndReenqueue(ProducerBatch bigBatch) {
             // We treat the newly split batches as if they are not even tried.
             synchronized (partitionDequeue) {
                 partitionDequeue.addFirst(batch);
+                if (transactionManager != null) {
+                    // We should track the newly created batches since they already have assigned sequences.
+                    transactionManager.startTrackingBatch(batch);
+                    maybeEnsureQueueIsOrdered(partitionDequeue, batch);
+                }
             }
         }
         return numSplitBatches;
     }
 
+    private void maybeEnsureQueueIsOrdered(Deque<ProducerBatch> deque, ProducerBatch batch) {
+        if (transactionManager != null) {
+            // When we are requeing and have enabled idempotence, the reenqueued batch must always have a sequence.
+            if (batch.baseSequence() == RecordBatch.NO_SEQUENCE)
+                throw new IllegalStateException("Trying to reenqueue a batch which doesn't have a sequence even " +
+                        "though idempotence is enabled.");
+
+            if (batch.baseSequence() != transactionManager.nextBatchBySequence(batch.topicPartition).baseSequence()) {
+                // The head of the queues have diverged. This means that the incoming batch should be placed somewhere further back.
+                // We need to find the right place for the incoming batch and insert it there.
+                // We will only enter this branch if we have multiple inflights sent to different brokers, perhaps
+                // because a leadership change occurred in between the drains. In this scenario, responses can come
+                // back out of order, requiring us to re order the batches ourselves rather than relying on the
+                // implicit ordering guarantees of the network client which are only on a per connection basis.
+
+                ProducerBatch incomingBatch = deque.pollFirst();
+                List<ProducerBatch> orderedBatches = new ArrayList<>();
+                while (deque.peekFirst() != null && deque.peekFirst().hasSequence() && deque.peekFirst().baseSequence() < incomingBatch.baseSequence())
```

Comment:
This is a good point. If the producer id changes, that means the transaction manager will stop tracking inflight batches for the previous producer id. So I updated the code to only try reordering if there are actually in flight batches by the current producer id. 

Once we have batches by the current producer id, we know that that all batches by the previous producer id are no longer in play, because that is the only way to get out of the unresolved state and start draining new batches.